middleware/rate_limiter_test.go

```diff
@@ -0,0 +1,230 @@
+package middleware
+
+import (
+	"github.com/labstack/echo/v4"
+	"github.com/stretchr/testify/assert"
+	"net/http"
+	"net/http/httptest"
+	"testing"
+)
+
+func TestRateLimiter(t *testing.T) {
+	e := echo.New()
+
+	handler := func(c echo.Context) error {
+		return c.String(http.StatusOK, "test")
+	}
+
+	{
+		var inMemoryStore = new(RateLimiterMemoryStore)
+		inMemoryStore.rate = 1
+		inMemoryStore.burst = 3
+
+		assert.Panics(t, func() {
+			RateLimiter(nil)
+		})
+
+		assert.NotPanics(t, func() {
+			RateLimiter(inMemoryStore)
+		})
+	}
+
+	{
+		var skipped bool
+		var inMemoryStore = new(RateLimiterMemoryStore)
+		inMemoryStore.rate = 1
+		inMemoryStore.burst = 3
+
+		req := httptest.NewRequest(http.MethodGet, "/", nil)
+		req.Header.Add(echo.HeaderXRealIP, "127.0.0.1")
+
+		rec := httptest.NewRecorder()
+
+		c := e.NewContext(req, rec)
+
+		mw := RateLimiterWithConfig(RateLimiterConfig{
+			Skipper: func(c echo.Context) bool {
+				skipped = true
+				return true
+			},
+			Store: inMemoryStore,
+			SourceFunc: func(ctx echo.Context) string {
+				return "127.0.0.1"
+			},
+		})
+
+		_ = mw(handler)(c)
+
+		assert.Equal(t, true, skipped)
+	}
+
+	{
+		var beforeRan bool
+		var inMemoryStore = new(RateLimiterMemoryStore)
+		inMemoryStore.rate = 1
+		inMemoryStore.burst = 3
+
+		req := httptest.NewRequest(http.MethodGet, "/", nil)
+		req.Header.Add(echo.HeaderXRealIP, "127.0.0.1")
+
+		rec := httptest.NewRecorder()
+
+		c := e.NewContext(req, rec)
+
+		mw := RateLimiterWithConfig(RateLimiterConfig{
+			BeforeFunc: func(c echo.Context) {
+				beforeRan = true
+			},
+			Store: inMemoryStore,
+			SourceFunc: func(ctx echo.Context) string {
+				return "127.0.0.1"
+			},
+		})
+
+		_ = mw(handler)(c)
+
+		assert.Equal(t, true, beforeRan)
+	}
+
+	var inMemoryStore = new(RateLimiterMemoryStore)
+	inMemoryStore.rate = 1
+	inMemoryStore.burst = 3
+
+	testCases := []struct {
+		id   string
+		code int
+	}{
+		{"127.0.0.1", 200},
+		{"127.0.0.1", 200},
+		{"127.0.0.1", 200},
+		{"127.0.0.1", 429},
+		{"127.0.0.1", 429},
+		{"127.0.0.1", 429},
+		{"127.0.0.1", 429},
```

Comment:
Guess we should use http.StatusTooManyRequests for the tests too for readability.

Reply:
Sounds sane to me. I'll add it right away